exec_test.go

```diff
@@ -247,30 +263,233 @@ func TestCommandStartWait(t *testing.T) {
+			require.Error(t, err, "Unexpected success calling (*Cmd).Wait before (*Cmd).Start")
 
 			cmd.Stdin = 0
-			cmd.Stderr = 0
-			cmd.Stdout = 0
-			err := cmd.Start()
+			err = cmd.Start()
 
-			// AFTER_START block
+			// AfterStart block
 			if tc.cancelOn == AfterStart {
 				cancel()
 			}
 			if requireErrors(t, tc, AfterStart, err) {
 				return
 			}
 
+			if stdout != nil {
+				text, err := stdout.ReadString('\n')
+				if err != nil {
+					require.ErrorIs(t, err, io.EOF, "Unexpected failure reading from StdoutPipe")
+				}
+				assert.Equal(t, tc.wantStdout, text, "Mismatch in piped stdout")
+			}
+
+			if stderr != nil {
+				text, err := stderr.ReadString('\n')
+				if err != nil {
+					require.ErrorIs(t, err, io.EOF, "Unexpected failure reading from StderrPipe")
+				}
+				assert.Equal(t, tc.wantStderr, text, "Mismatch in piped stderr")
+			}
+
+			err = cmd.Start()
+			require.Error(t, err, "Unexpected success calling (*Cmd).Start twice")
+
+			_, err = cmd.StdoutPipe()
+			require.Error(t, err, "Unexpected success calling (*Cmd).StdoutPipe after (*Cmd).Start")
+
+			_, err = cmd.StderrPipe()
+			require.Error(t, err, "Unexpected success calling (*Cmd).StderrPipe after (*Cmd).Start")
+
 			err = cmd.Wait()
 
-			// AFTER_WAIT block
+			// AfterWait block
 			if requireErrors(t, tc, AfterWait, err) {
 				return
 			}
+
+			err = cmd.Wait()
+			require.Error(t, err, "Unexpected success calling (*Cmd).Wait twice")
+		})
+	}
+}
+
+func TestCommandOutPipes(t *testing.T) {
+	d := newTestDistro(t, jammyRootFs)
+
+	testCases := map[string]struct {
+		cmd    string
+		stdout bool
+		stderr bool
+
+		wantRead string
+	}{
+		"all discarded":           {},
+		"piped stdout":            {stdout: true, wantRead: "Hello stdout\n"},
+		"piped stderr":            {stderr: true, wantRead: "Hello stderr\n"},
+		"piped stdout and stderr": {stdout: true, stderr: true, wantRead: "Hello stdout\nHello stderr\n"},
+	}
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			cmd := d.Command(context.Background(), "echo 'Hello stdout' >& 1 && sleep 1 && echo 'Hello stderr' >& 2")
+
+			var buff bytes.Buffer
+			if tc.stdout {
+				cmd.Stdout = &buff
+			}
+			if tc.stderr {
+				cmd.Stderr = &buff
+			}
+
+			err := cmd.Run()
+			require.NoError(t, err, "Did not expect an error when launching command")
+
+			require.NoError(t, err, "Did not expect read from pipe to return an error")
+			require.Equal(t, tc.wantRead, buff.String())
+		})
+	}
+}
+
+func TestCommandOutput(t *testing.T) {
+	realDistro := newTestDistro(t, jammyRootFs)
+	fakeDistro := wsl.Distro{Name: UniqueDistroName(t)}
+	wrongDistro := wsl.Distro{Name: UniqueDistroName(t) + "--IHaveA\x00NullChar!"}
+
+	testCases := map[string]struct {
+		distro       *wsl.Distro
+		cmd          string
+		presetStdout io.Writer
+
+		wantStdout      string
+		errorWanted     bool
+		exitErrorWanted bool
+		// Only relevant if exitErrorWanted==true
+		wantExitCode uint32
```

Comment:
I think I see what you did that, it’s because of bool vs value, I still find it confusing personnally though :)

Reply:
Yeah I suppose I did it subcontiously. Still, I changed it to `wantThing` for consistency.